hazelcast/src/main/java/com/hazelcast/map/impl/MapEntries.java

```diff
@@ -35,31 +35,139 @@
  */
 public final class MapEntries implements IdentifiedDataSerializable {
 
+    private static interface ValueWrapper {
```

Comment:
I leave it to @pveentjer to provide the benchmarks, given the hazelcast team would have seen far more where this is a problem, relative to my own experiences as a user. 

I suggest as an alternative retaining two `Lists` within MapEntries, but changing the type param of values to Object, then adjusting the rest of the PR as needed. I realize that this isn't as clean as you would like, but this may well be the compromise that would _both_ avoid GC litter _and_ defer serialization.